### DISCUSSION TITLE
Zygisk not working? Is it supported?

### DISCUSSION THREAD

**Discussion by alphapuggle** [chenxiaolong/avbroot#266] [General]
Just installed GrapheneOS (not doing it for security before anyone says it, I wanted the sandboxed Google Play), and have tried with both Kitsune Mask and normal Magisk (Kitsune supposedly fixed this, but it didn't have any change).

The issue is that Zygisk will always say that it's not enabled, and the Zygisk toggle says "Reboot to apply changes"

I don't know if Zygisk is *supposed* to work, I don't see anything mentioned about it in the README or in issues or other discussions. If any logs would help (they would, I just don't know which ones) let me know.

**Comment by pascallj** [comment]
You might wanna read [this](https://github.com/chenxiaolong/avbroot/issues/213).

Seems like Magisk Delta was the solution there. However the rest of the issue is interesting as well regarding GrapheneOS. Which you might reconsider if root is important to you.

**Comment by pixincreate** [comment]
I've been maintaining a version of Magisk that specifically fixes Zygisk on GrapheneOS (since I use it personally) which will always be in sync with upstream version of Magisk.

You might want to check that out here if interested: https://github.com/pixincreate/Magisk

Rooting is discouraged given how greatly it reduces security unlike other unix based systems and only use it only of you know what you're doing.

**Reply by ROBOT0-VT** [reply]
I've started using your fork along with AVBRoot for my own auto-update script, it's literally the same fork that I was going to make :D So thank you!

**Reply by ubergeek77** [reply]
Hey @pixincreate , I see you archived your repo yesterday.

I was planning on adopting your patches, so I'm curious, did the newest Magisk Canary break things too much, making it impossible to maintain? Or are your patches still working fine for now?

**Reply by pixincreate** [reply]
No, it works. It is just that I got too frustrated with hiding root.  
None of the root hiding solutions work as expected (I found a need to actually hide it this time). Especially `Shamiko`. I tried going around that and couldn't get working. Maintenance wise, I have had no issues with it so far.

**Reply by ubergeek77** [reply]
Oh, thanks for letting me know. I'm glad to hear it works! And really good job with the pipeline, it's a clean way to automate updates, I learned a few tricks from you by reading it :)

I hope Magisk/GrapheneOS doesn't break your patch for at least a little while. Your fork was the only one I could find that had up to date patches for Zygisk+GrapheneOS. Looks like Kitsune hasn't been updated in about 6 months.

Sorry to see you put it down, but thanks for making it in the first place!

**Reply by pixincreate** [reply]
I flashed my phone and locked the bootloader with plans to never come back to root. I'm in fact happy to continue maintaining it if people are really using it and finding it useful.  

I also have plans to modify BCR to make it support Shizuku for automated call recording (and possibly raise a [PR](https://github.com/chenxiaolong/BCR) ) as I learnt that it is [possible](https://github.com/chenxiaolong/BCR/issues/278) given that additional permissions do not require system level access. Not sure how would that go.

Update: Unarchived. I personally would not be able to test it going forward.

**Reply by ubergeek77** [reply]
I compile my own version of GrapheneOS and integrate Magisk into the build system. Even though I can make my own tweaks and have a lot of control, I still need Magisk for stuff like Neo Backup. Zygisk isn't perfect , but it's helpful when it does work against root detection, mostly just a few games. So personally, I would really appreciate it if you kept maintaining it :)

At least long enough for GrapheneOS to finish their system backup app they keep promising. That's one of the last pieces I think I'd need before not needing root anymore.

As for BCR and Shizuku, wow - that's an insane discovery you've made. I think myself and a ***ton*** of people would be extremely happy if you made rootless automatic call recording possible. I'm glad GrapheneOS has call recording officially, but it not being automatic is really annoying. I've missed out on some recorded calls I really needed, all because I forgot to enable the recorder.

Even if I needed to add some special permissions for the BCR app, that's totally fine, since I build my own GrapheneOS from source.

Anyway, you're doing some great work here, thank you very much!

**Reply by pixincreate** [reply]
My guess is that, since you build a version of your own, you could bake BCR within the ROM itself? 

> At least long enough for GrapheneOS to finish their system backup app they keep promising. That's one of the last pieces I think I'd need before not needing root anymore.

I tried it yesterday, it just works for some reason. There are hiccups here and there w.r.t apps that failed restore, spam servers every now then "trying" to restore in the background. And snapchat blocked my account for the same. Texting apps like Signal, WhatsApp and Telegram did not get restored. So did Proton apps. Most of the settings (especially quick setting tiles) got restored perfectly and I loved it. I just to had to reset network settings once to get my Payment app working.

> As for BCR and Shizuku, wow - that's an insane discovery you've made.

I was looking for the same until I accidentally tumbled on that thread, not a discovery though, lol. I'm currently resorting to use `Cube ACR` for automated call recording until I try messing around BCR. 

Fingers crossed.🤞

**Reply by ubergeek77** [reply]
> My guess is that, since you build a version of your own, you could bake BCR within the ROM itself?

Yep! Putting app packages together is pretty simple. The only complicated part is making sure I define the permissions it needs, and that BCR itself can handle those permissions. If that's sorted, then it should all work fine. I may not even need Shizuku or root, if BCR can be patched to work as a system app in this way.

I didn't have quite the positive experience with Seedvault as you, but at least a replacement is being developed regardless, right? :p

**Reply by pixincreate** [reply]
> The only complicated part is making sure I define the permissions it needs, and that BCR itself can handle those permissions.

I saw it only required 2 permission for recording calls. May be the developer might have more insight on that.  

> I didn't have quite the positive experience with Seedvault as you, but at least a replacement is being developed regardless, right?

I never had either, but I was forced to use it yesterday I wanted a seamless transition. I used USB flash drive for restore BTW.

**Reply by chenxiaolong** [reply]
> Even if I needed to add some special permissions for the BCR app, that's totally fine, since I build my own GrapheneOS from source.

If you're building GrapheneOS from source, the only files you need to add for BCR are `/system/priv-app/com.chiller3.bcr/app-release.apk` and `/system/etc/permissions/privapp-permissions-com.chiller3.bcr.xml`. Everything aside from the functionality to read the other party's name from the call log will work out of the box.

The call log functionality is not very useful on GrapheneOS anyway. Its purpose is for taking advantage of dialers that perform reverse lookups, like Google Dialer on the stock OS.

<details>
<summary>But is is possible to set up if you want it:</summary>

Ensure you're building GrapheneOS as a `userdebug` build so that `su` is available and then add `/system/etc/init/bcr_remove_hard_restrictions.rc`:

```
service bcr_remove_hard_restrictions /system/bin/app_process / com.chiller3.bcr.standalone.RemoveHardRestrictionsKt
    oneshot
    class main
    seclabel u:r:su:s0
    setenv CLASSPATH /system/priv-app/com.chiller3.bcr/app-release.apk
```

(For a normal `user` build, you would instead have to modify the something in `frameworks/base/` so that it grants `PackageManager.FLAG_PERMISSION_RESTRICTION_SYSTEM_EXEMPT` to BCR.)
</details>

This is actually what I personally do. I no longer use Magisk and instead, integrate BCR and some other things directly into the OS: https://github.com/chenxiaolong/my-avbroot-setup. I opted to use avbroot + [afsr](https://github.com/chenxiaolong/afsr) for this because the iteration time is so much shorter than rebuilding GrapheneOS when I'm developing things.

My usage is probably different from most people though. I don't use any root apps and only use root for reverse engineering. GrapheneOS userdebug builds (with `ro.adb.secure=1`) includes AOSP `su` binary that's only accessible via adb and that's good enough for me.

---

Regarding using BCR on an unrooted system, I wrote about it a bit here: https://github.com/chenxiaolong/BCR/issues/278#issuecomment-1925914290, but it would probably be painful to implement.

If there's some way to use `InCallService` with Shizuku, I'd be open to accepting a PR. But I really do not want to use Android's legacy call detection mechanism.

**Reply by chenxiaolong** [reply]
I just merged a new feature in BCR that makes a potential rootless implementation even harder to implement (sorry): https://github.com/chenxiaolong/BCR/pull/573

One of GrapheneOS' good security features is auto-rebooting to bring the device back to the BFU state. But that increases the likelihood of answering calls before the device is initially unlocked, which previously couldn't be recorded. BCR now supports recording these calls.

**Reply by ubergeek77** [reply]
Thanks for the heads up @chenxiaolong ! Skimming that PR, I don't really see anything that jumps out as making this harder to implement rootlessly.

At least, baking this into a from-source AOSP build with all features intact looks straightforward enough. You use all the standard Android permissions, Intents, Content Resolvers instead of hardcoded paths, etc. Of course using this rootlessly on a stock ROM is definitely out of the question.

I'll need to dig up one of my old phones and try this out one day, but at least people like me should in theory be able to use this as-is in an AOSP build.

Thanks for your amazing work on this!

**Reply by chenxiaolong** [reply]
No problem!

> Of course using this rootlessly on a stock ROM is definitely out of the question.

I no longer use the stock OS, but this setup is something I tested out. Adding the necessary files to `system` with afsr and then patching/signing the OTA with avbroot (`--rootless` mode) works with the stock OS too. (Though this won't really help for folks who want to run a completely unmodified stock OS.)

**Reply by pixincreate** [reply]
> I just merged a new feature in BCR that makes a potential rootless implementation even harder to implement (sorry): [chenxiaolong/BCR#573](https://github.com/chenxiaolong/BCR/pull/573)
> 
> One of GrapheneOS' good security features is auto-rebooting to bring the device back to the BFU state. But that increases the likelihood of answering calls before the device is initially unlocked, which previously couldn't be recorded. BCR now supports recording these calls.

So `FOREGROUND_SERVICE_SPECIAL_USE` requires special system level permission? Asking so because `RECEIVE_BOOT_COMPLETED` is added in my version of Shizuku for supporting auto restart and it does not require any special permission. 

This feature does makes a lot of sense. I think I should stop trying to get BCR working on rootless devices.  
I haven't given a shot onto it yet (But I would want to try that on weekend).

I'm thinking of moving that feature behind a feature flag such that, if the device does not have root access, then call recording during BFU should not work.

And yes, if that's really useless, then I won't even try.

**Reply by chenxiaolong** [reply]
> So `FOREGROUND_SERVICE_SPECIAL_USE` requires special system level permission?

Nope, no special restrictions on that. It's available to all apps.

> I'm thinking of moving that feature behind a feature flag such that, if the device does not have root access, then call recording during BFU should not work.

Yeah, when I said the new feature might make things harder to implement, this is the sort of thing I was referring to. (One more thing that would need to be disabled/removed for rootless operation.)

**Reply by pixincreate** [reply]
Oh, I see. I'll give it a shot then. And yes, I'm an "ultra beginner" in Android 😅

**Reply by e-t-l** [reply]
> I compile my own version of GrapheneOS and integrate Magisk into the build system

What's the benefit of doing that instead of patching Magisk into OTAs with avbroot?

**Reply by ubergeek77** [reply]
There isn't one, I just like building GrapheneOS from source and applying my own UI tweaks.

**Reply by e-t-l** [reply]
> applying my own UI tweaks

Like what? I'm intrigued

**Comment by chenxiaolong** [comment]
pascallj and pixincreate already gave good answers for Magisk forks to try, but I wanted to point out something related to this:

> I don't see anything mentioned about it in the README or in issues or other discussions

All of the Zygisk components are executed after most of the system is already booted. They are not part of the boot image and there's nothing in avbroot that would affect Zygisk specifically.
